include/deal.II/fe/fe_enriched.h

```diff
@@ -0,0 +1,585 @@
+  Year                     = {1997},
+  Number                   = {4},
+  Pages                    = {727--758},
+  Volume                   = {40},
+}
+ * @endcode
+ *
+ * @ingroup fe
+ *
+ * @author Denis Davydov, 2016.
+ */
+template <int dim, int spacedim=dim>
+class FE_Enriched : public FiniteElement<dim, spacedim>
+{
+public:
+
+  /**
+   * Constructor which takes base FiniteElement @p fe_base and the enrichment
+   * FiniteElement @p fe_enriched which will be multiplied by the @p enrichment_function.
+   *
+   * In case @p fe_enriched is other than FE_Nothing, the lifetime of the
+   * @p enrichment_function must be at least as long as the FE_Enriched object.
+   */
+  FE_Enriched (const FiniteElement<dim,spacedim> &fe_base,
+               const FiniteElement<dim,spacedim> &fe_enriched,
+               const Function<spacedim>      *enrichment_function);
+
+  /**
+   * Constructor which only wraps the base FE @p fe_base.
+   * As for the enriched finite element space, FE_Nothing is used.
+   * Continuity constraints will be automatically generated when
+   * this non-enriched element is used in conjunction with enriched finite element
+   * within the hp::DoFHandler.
+   *
+   * See the discussion in the class documentation.
+   */
+  FE_Enriched (const FiniteElement<dim,spacedim> &fe_base);
+
+  /**
+   * Constructor which takes pointer to the base FiniteElement @p fe_base and
+   * a vector of enriched FiniteElement's @p fe_enriched . @p fe_enriched[i]
+   * finite element will be enriched with functions in @p functions[i].
+   *
+   * This is the most general public constructor which also allows to have
+   * different enrichment functions in different disjoint parts of the domain.
+   * To that end the last argument provides an association of cell iterator
+   * to a Function. This is done to simplify the usage of this class when the
+   * number of disjoint domains with different functions is more than a few.
+   * Otherwise one would have to use different instance of this class for each
+   * disjoint enriched domain.
+   *
+   * @note When using the same finite element for enrichment with N
+   * different functions, it is advised to have the second argument of size 1
+   * and the last argument of size 1 x N. The same can be achieved by providing
+   * N equivalent enrichment elements while keeping the last argument of size
+   * N x 1. However this will be much more computationally expensive.
+   *
+   * @note When using different enrichment functions on disjoint domains, no
+   * checks are done by this class that the domains are actually disjoint.
+   */
+  FE_Enriched (const FiniteElement<dim,spacedim> *fe_base,
+               const std::vector<const FiniteElement<dim,spacedim> * > &fe_enriched,
+               const std::vector<std::vector<std::function<const Function<spacedim> *(const typename Triangulation<dim, spacedim>::cell_iterator &) > > > &functions);
```

Comment:
the signature of the last argument is quite complex. I think it would be useful to have an example that explains how to use it.

Or did you intend to write a little tutorial program, maybe, that shows this? That would be even better!